integrations/transportation/tfl.js

```diff
@@ -0,0 +1,173 @@
+          distanceKilometers: journey.TransactionType === 'Bus' ? 3.140 : null, // a floating point that represents the amount of kilometers traveled (https://www.whatdotheyknow.com/request/bus_passenger_journey_times)
+          activityType: ACTIVITY_TYPE_TRANSPORTATION,
+          transportationMode: TRANSPORTATION_MODE_PUBLIC_TRANSPORT, // a variable (from definitions.js) that represents the transportation mode
+          carrier: 'Transport For London', // (optional) a string that represents the transportation company
+          departureStation: journey.TransactionType === 'Bus' ? 'Bus' : journey.StartLocation, // (for other travel types) a string that represents the original starting point
+          destinationStation: journey.TransactionType === 'Bus' ? 'Bus' : journey.EndLocation, // (for other travel types) a string that represents the final destination
+        });
+      }
+    });
+  });
+  return activities;
+}
+
+async function connect(requestLogin) {
+  const { username, password } = await requestLogin();
+  if (!(password || '').length) {
+    throw new ValidationError('Password cannot be empty');
+  }
+
+  const postBody = {
+    username,
+    password,
+    AppId: '9C9C6B6C-A025-493E-8F39-3A6D57C7ACAB', // TODO what do we do with this?
+  };
+
+  const loginResponse = await fetch(LOGIN_PATH, {
+    method: 'post',
+    headers: {
+      'Content-type': 'application/json',
+    },
+    body: JSON.stringify(postBody),
+  })
+    .then(response => response.json())
+    .catch((e) => {
+      throw new HTTPError(e);
+    });
+
+  if (!loginResponse.SecurityToken) throw new HTTPError('Login failed');
+
+  // console.log(loginResponse);
+
+  const apiTokenResponse = await fetch(`${BASE_PATH}/APITokens`, {
+    method: 'get',
+    headers: {
+      code: loginResponse.SecurityToken,
+      grant_type: 'authorization_code',
+    },
+  })
+    .then(response => response.json())
+    .catch((e) => {
+      throw new HTTPError(e);
+    });
+
+  // console.log(apiTokenResponse);
+
+  return {
+    securityToken: loginResponse.SecurityToken,
+    refreshToken: apiTokenResponse.refresh_token,
+    accessToken: apiTokenResponse.access_token,
+  };
+}
+
+async function getOneWeeksOysterData(accessToken, oysterCardNumber, startMoment, endMoment) {
+  return fetch(`${BASE_PATH}/Cards/Oyster/Journeys?startDate=${startMoment.format(moment.HTML5_FMT.DATE)}&endDate=${endMoment.format(moment.HTML5_FMT.DATE)}`, {
+    method: 'get',
+    headers: {
+      'x-zumo-auth': accessToken,
+      oystercardnumber: oysterCardNumber,
+    },
+  })
+    .catch((e) => {
+      throw new HTTPError(e);
+    });
+}
+
+async function collect(state) {
+  // Get access token using the refresh token
+  const refreshTokenResponse = await fetch(`${BASE_PATH}/APITokens/RefreshToken`, {
```

Comment:
I'd suggest testing against the token expiry time before requesting a new refreshToken, no need to do it on every request. See Oath2Manager for an example

- https://github.com/tmrowco/tmrowapp-contrib/blob/d5ce1be4e9919412ca2015c4df18f1c308a60e65/integrations/authentication/OAuth2Manager.js#L53
- https://github.com/tmrowco/tmrowapp-contrib/blob/d5ce1be4e9919412ca2015c4df18f1c308a60e65/integrations/authentication/OAuth2Manager.js#L125

Reply:
Good point, will use the same logic